core/src/main/scala/io/github/scalaquest/core/Game.scala

```diff
@@ -1,23 +1,31 @@
 package io.github.scalaquest.core
 
 import io.github.scalaquest.core.model.{Message, Model}
-import io.github.scalaquest.core.pipeline.Pipeline
+import io.github.scalaquest.core.pipeline.Pipeline.PipelineBuilder
 
-trait Game[S <: Model#State] {
-  def send(input: String)(state: S): Either[String, S]
-}
+abstract class Game[M <: Model](val model: M) {
 
-trait GameTemplate[S <: Model#State] extends Game[S] {
-  def pipelineFactory(state: S): Pipeline[S]
-  override def send(input: String)(state: S): Either[String, S] = pipelineFactory(state) run input
+  def send(input: String)(state: model.S): Either[String, model.S]
 }
 
 trait MessagePusher extends (Seq[Message] => Seq[String])
 
 object Game {
 
-  def apply[S <: Model#State](_pipelineFactory: S => Pipeline[S]): Game[S] =
-    new GameTemplate[S] {
-      override def pipelineFactory(state: S): Pipeline[S] = _pipelineFactory(state)
+  class GameFromModel[M <: Model](val model: M) {
+
+    abstract class GameWithPipeline extends Game[model.type](model) {
+
+      def pipelineFactory: PipelineBuilder[model.S, model.type]
+
+      override def send(input: String)(state: model.S): Either[String, model.S] = pipelineFactory(state) run input
     }
+
+    def withPipelineBuilder(pipelineBuilder: PipelineBuilder[model.S, model.type]): Game[model.type] =
+      new GameWithPipeline {
+        override def pipelineFactory: PipelineBuilder[model.S, model.type] = pipelineBuilder
```

Comment:
This line gives this error with IntelliJ, even if the code compiles successfully. Is this only an error from IntelliJ?


![Screenshot 2021-01-03 at 16 15 31](https://user-images.githubusercontent.com/22854792/103482120-32303d00-4ddf-11eb-837e-9493ec6ae8c8.png)

Reply:
I don't know honestly. @francescogorini and I get it too, but the compiler doesn't even report it as a warning, so I'm kind of confident it's an highlighting error of the IDE (I hope).

Reply:
The code seems good, actually. I think we've passed IntelliJ limits', but as far as it is valid Scala code, accepted by the compiler, this is ok